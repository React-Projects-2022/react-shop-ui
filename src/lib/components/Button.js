import React from "react";
import PropTypes from "prop-types";
const Button = ({ text, className, handleAction, params }) => (
  <button
    type="button"
    className={className}
    onClick={() => handleAction(params)}
  >
    {text}
  </button>
);

export { Button };

Button.propTypes = {
  handleAction: PropTypes.func.isRequired,
};
