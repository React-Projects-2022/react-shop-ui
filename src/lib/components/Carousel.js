import React from "react";
import PropTypes, { object } from "prop-types";
import "./../styles/carousel.css";
const Carousel = ({ carousel }) => (
  <header>
    <div
      id="carouselExampleCaptions"
      className="carousel slide"
      data-bs-ride="carousel"
      data-bs-interval="3000"
    >
      <div className="carousel-indicators">
        {carousel.map((item, index) => (
          <button
            key={item.id ? item.id : index}
            type="button"
            data-bs-target="#carouselExampleCaptions"
            data-bs-slide-to={index}
            className={index === 0 ? "active" : ""}
            aria-current="true"
            aria-label={"Slide ".concat(index + 1)}
          ></button>
        ))}
      </div>
      <div className="carousel-inner">
        {carousel.map((item, index) => (
          <div
            key={item.id ? item.id : index}
            className={index === 0 ? "carousel-item active" : "carousel-item"}
            style={{
              backgroundImage: `url("${item.background}")`,
            }}
          >
            <div className="carousel-caption">
              <h5>{item.title}</h5>
              <p>{item.description}</p>
            </div>
          </div>
        ))}
      </div>
      <button
        className="carousel-control-prev"
        type="button"
        data-bs-target="#carouselExampleCaptions"
        data-bs-slide="prev"
      >
        <span className="carousel-control-prev-icon" aria-hidden="true"></span>
        <span className="visually-hidden">Previous</span>
      </button>
      <button
        className="carousel-control-next"
        type="button"
        data-bs-target="#carouselExampleCaptions"
        data-bs-slide="next"
      >
        <span className="carousel-control-next-icon" aria-hidden="true"></span>
        <span className="visually-hidden">Next</span>
      </button>
    </div>
  </header>
);

export { Carousel };

Carousel.propTypes = {
  carousel: PropTypes.array.isRequired
};


/**
 * {
          id: item.id,
          title: item.product.name,
          description: item.platform.name,
          url: window.location.origin.concat(`/games/details/${item.id}`),
          background: item.product.img,
        }
 */