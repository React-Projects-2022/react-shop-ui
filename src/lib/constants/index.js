import { CURRENCIES_SYMBOL, CURRENCY_LIST } from "./currencies";
const CONSTANTS = { CURRENCIES_SYMBOL, CURRENCY_LIST }
export { CONSTANTS };