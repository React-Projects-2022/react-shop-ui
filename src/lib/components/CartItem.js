import { useState } from "react";
import PropTypes from "prop-types";
import { CONSTANTS } from "./../constants";
import { QuantitySelector } from "./QuantitySelector";
export const CartItem = ({
  product,
  selectMoney,
  updateValue,
  clearItem,
  navigateTo
}) => {
  const [subtotal, setSubtotal] = useState(product.qty * product.price)
  return (
    <tr>
      <td>
        <div className="product-item" key={product.id} onClick={() => navigateTo("games/details", product.id)}>
          <span className="product-thumb">
            <img src={product.img} alt={product.name} />
          </span>
          <div className="product-info">
            <h4 className="product-title">{product.name}</h4>
            <span>
              <em>Price:</em> {CONSTANTS.CURRENCIES_SYMBOL[selectMoney]}{" "}
              {product.price}
            </span>
            <span>
              <em>Platform:</em> {product.description}
            </span>
          </div>
        </div>
      </td>
      <td className="text-center">
        <QuantitySelector
          stock={product.stock}
          updateValue={(counter) => {
            updateValue(counter, product.id);
            product.qty = counter;
            setSubtotal(product.price * product.qty);
          }}
          count={product.qty}
        />
      </td>
      <td className="text-center text-lg text-medium">
        {CONSTANTS.CURRENCIES_SYMBOL[selectMoney]}{" "}
        {Number.parseFloat(subtotal).toFixed(2)}
      </td>
      <td className="text-center">
        <button
          onClick={() => clearItem(product.id)}
          className="remove-from-cart btn btn-danger"
          title={ `Delete ${product.name} product in cart`}
        >
          <i className="fa fa-trash"></i>
        </button>
      </td>
    </tr>
  );
};

CartItem.propTypes = {
  product: PropTypes.object.isRequired,
  selectMoney: PropTypes.string.isRequired,
  updateValue: PropTypes.func.isRequired,
  clearItem: PropTypes.func.isRequired,
  navigateTo: PropTypes.func.isRequired
};