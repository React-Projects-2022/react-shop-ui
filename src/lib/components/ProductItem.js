import "./../styles/product-item.css";
import { Rating } from "./Rating";
import PropTypes from "prop-types";
import { Button } from "./Button";
import { CONSTANTS } from "../constants";
export const ProducItem = ({
  item,
  showDetails,
  addCart,
  showDesc,
  selectMoney,
}) => {
  // https://bbbootstrap.com/snippets/product-card-template-hover-effect-and-animation-23224168
  return (
    <>
      {item ? (
        <div className="card">
          <div className="heart">
            <i className="fa fa-heart"></i>
          </div>
          <div className="top-div">
            <div className="border">
              <img src={item.product.img} alt="" />
            </div>
            <span>
              {CONSTANTS.CURRENCIES_SYMBOL[selectMoney]} {item.price}
            </span>
          </div>
          <div className="bottom-div">
            <h5 className="truncate-one-line">{item.product.name}</h5>
            {showDesc ? <p>{item.platform.name}</p> : <br />}
            <Rating
              max={5}
              ratingItem={{
                value: item.product.rating.value,
                count: item.product.rating.count,
              }}
            />
          </div>
          <div className="last-section text-center">
            <div className="buttons">
              <Button
                params={item.id}
                text={<i className="fas fa-info-circle"></i>}
                handleAction={showDetails}
              />
              &nbsp;
              <Button
                params={item.id}
                text={<i className="fas fa-cart-plus"></i>}
                handleAction={addCart}
              />
            </div>
          </div>
        </div>
      ) : (
        ""
      )}
    </>
  );
};

ProducItem.propTypes = {
  item: PropTypes.object.isRequired,
  showDetails: PropTypes.func.isRequired,
  addCart: PropTypes.func.isRequired,
  selectMoney: PropTypes.oneOf(Object.values(CONSTANTS.CURRENCY_LIST)),
};
