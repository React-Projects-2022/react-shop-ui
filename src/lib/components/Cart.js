import PropTypes from "prop-types";
import { CONSTANTS } from "../constants";
import { CartItem } from "./CartItem";
import { useCart } from "../hooks/useCart";
import "./../styles/cart.css";
export const Cart = ({ navigateTo }) => {
  const {
    cart: cartData,
    clearCart,
    clearItem,
    updateValue,
    updateCart,
    total,
  } = useCart();

  const selectMoney = CONSTANTS.CURRENCY_LIST.EURO;

  return (
    <>
      {Object.keys(cartData).length > 0 && cartData.products.length > 0 ? (
        <>
          <div className="table-responsive shopping-cart">
            <table className="table">
              <thead>
                <tr>
                  <th>Product Name</th>
                  <th className="text-center">Quantity</th>
                  <th className="text-center">Subtotal</th>
                  <th className="text-center">
                    <button
                      className="btn btn-sm btn-outline-danger"
                      onClick={clearCart}
                    >
                      <i className="fas fa-trash-alt"></i> Clear Cart
                    </button>
                  </th>
                </tr>
              </thead>
              <tbody>
                {cartData.products.map((product) => (
                  <CartItem
                    product={product}
                    key={product.id}
                    selectMoney={selectMoney}
                    updateValue={updateValue}
                    clearItem={clearItem}
                    navigateTo={navigateTo}
                  />
                ))}
              </tbody>
            </table>
          </div>
          <div className="shopping-cart-footer">
            <div className="column text-lg">
              To Pay:{" "}
              <span className="text-primary price-total">
                {" "}
                {CONSTANTS.CURRENCIES_SYMBOL[selectMoney]}{" "}
                {Number.parseFloat(total).toFixed(2)}{" "}
              </span>
            </div>
          </div>
          <div className="shopping-cart-footer">
            <div className="column">
              <button
                className="btn btn-outline-secondary"
                onClick={() => navigateTo("")}
              >
                <i className="fas fa-home"></i>&nbsp;Back to Shopping
              </button>
            </div>
            <div className="column">
              <button className="btn btn-danger" onClick={clearCart}>
                <i className="fas fa-trash-alt"></i> Clear Cart
              </button>
              <button className="btn btn-success" href="#">
                Checkout
              </button>
            </div>
          </div>
        </>
      ) : (
        <>
          <span>
            No tienes elementos guardados en el carrito de compra. A침치delos
            desde los productos con la opci칩n de A침adir a la cesta
          </span>
          <button onClick={() => updateCart()}>Load cart</button>
        </>
      )}
    </>
  );
};

Cart.propTypes = {
  navigateTo: PropTypes.func.isRequired,
};