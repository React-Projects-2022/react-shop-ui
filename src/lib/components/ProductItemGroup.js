import { ProducItem } from "./ProductItem";
import PropTypes from "prop-types";
import { CONSTANTS } from "../constants";
const ProductItemGroup = ({
  title,
  products,
  showDetails,
  addCart,
  showDesc,
  selectMoney,
}) => (
  <>
    <h2>{title}</h2>
    <hr />
    <div className="row row-cols-1 row-cols-md-4 g-4">
      {products.map((item) => (
        <div key={item.id} className="col">
          <ProducItem
            item={item}
            showDetails={showDetails}
            addCart={addCart}
            showDesc={showDesc}
            selectMoney={selectMoney}
          />
        </div>
      ))}
    </div>
    <hr />
  </>
);

export { ProductItemGroup };

ProductItemGroup.propTypes = {
  title: PropTypes.string.isRequired,
  products: PropTypes.array.isRequired,
  showDetails: PropTypes.func.isRequired,
  addCart: PropTypes.func.isRequired,
  selectMoney: PropTypes.oneOf(Object.values(CONSTANTS.CURRENCY_LIST)),
};
