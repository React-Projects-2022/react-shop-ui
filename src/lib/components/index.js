import React from "react";
import { QuantitySelector } from "./components/QuantitySelector";
import { ProductItem } from "./components/ProductItem";
import { ProductItemGroup } from "./components/ProductItemGroup";
import { Rating } from "./components/Rating";
import { Button } from "./components/Button";
import { Carousel } from "./components/Carousel";
import { Cart } from "./components/Cart";

export {
  Button,
  Carousel,
  Cart,
  ProductItem,
  ProductItemGroup,
  Rating,
  QuantitySelector,
};
