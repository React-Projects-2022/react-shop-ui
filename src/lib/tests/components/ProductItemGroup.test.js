import React from "react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { ProductItem } from "./../../components/ProductItem";
import { CONSTANTS } from "../../constants";
import { ProductItemGroup } from "../../components/ProductItemGroup";
describe("Pruebas en ProductItem", () => {
  let wrapper;
  const text = `<i className="fas fa-info-circle"></i>`;
  const handleAction = jest.fn();
  const showDetails = jest.fn();
  const addCart = jest.fn();
  const selectMoney = CONSTANTS.CURRENCY_LIST.EURO;
  const showDesc = true;
  const products = [
    {
      id: "1222",
      price: 23,
      product: {
        img: "",
        name: "Super Mario",
        rating: {
          value: 1,
          count: 340,
        },
      },
      platform: {
        name: "Nintendo 64",
      },
    },
    {
      id: "123",
      price: 56,
      product: {
        img: "",
        name: "Tomb Raider",
        rating: {
          value: 1,
          count: 340,
        },
      },
      platform: {
        name: "Playstation",
      },
    },
  ];

  beforeEach(() => {
    wrapper = shallow(<ProductItemGroup title={"Juegos"} products={products} showDesc={showDesc} showDetails={showDetails}
        addCart={addCart} selectMoney={selectMoney}/>);
  });
  test("Debe de mostrar <ProductItemGroup /> correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Comprobar que tenemos dos productos", () => {
    expect(wrapper.find("ProductItem").length).toBe(2);
  });
});
/*


  test("Comprobar las propiedades de los botones", () => {
    const buttonComponent = wrapper.find("Button");
    expect(buttonComponent.at(0).prop("params")).toBe(item.id);
    expect(buttonComponent.at(1).prop("params")).toBe(item.id);
    expect(buttonComponent.at(0).prop("text")).toMatchObject(
      <i className="fas fa-info-circle"></i>
    );
    expect(buttonComponent.at(1).prop("text")).toMatchObject(
      <i className="fas fa-cart-plus"></i>
    );
    // expect(buttonComponent.at(0).prop("handleAction")).toBeDefined(showDetails);
    // expect(buttonComponent.at(1).prop("handleAction")).toBeDefined(addCart);
  });

  test("Comprobar elementos con los datos si son correctos o no", () => {
    const productNameContent = wrapper.find(".truncate-one-line").text();
    expect(productNameContent).toBe(item.product.name);

    const productDescriptionElement = wrapper.find(".description");
    showDesc
      ? expect(productDescriptionElement.text()).toBe(item.platform.name)
      : expect(productDescriptionElement).toMatchObject({});

    const priceContent = wrapper.find(".top-div span").text();
    expect(priceContent).toBe(`â‚¬ ${item.price}`);
  });

  test("Comprobar que seleccionamos correctamente las monedas para mostrar", () => {
    const currenciesValues = Object.values(CONSTANTS.CURRENCY_LIST);
    Object.keys(CONSTANTS.CURRENCY_LIST).map((__, index) => {
      checkIfMoneySymbolIsCorrect(
        currenciesValues[index],
        CONSTANTS.CURRENCIES_SYMBOL[currenciesValues[index]],
        item
      );
    });
  });
});

const checkIfMoneySymbolIsCorrect = (selectMoney, expectedSymbol, item) => {
  console.log(expectedSymbol, item, selectMoney);
  const product = shallow(
    <ProductItem
      item={item}
      addCart={jest.fn()}
      showDesc={false}
      showDetails={jest.fn()}
      selectMoney={selectMoney}
    />
  );
  const priceContent = product.find(".top-div span").text();
  expect(priceContent).toBe(`${expectedSymbol} ${item.price}`);

  expect(product).toMatchSnapshot();
};*/
