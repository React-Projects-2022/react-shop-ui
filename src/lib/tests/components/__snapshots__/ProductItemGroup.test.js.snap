// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en ProductItem Debe de mostrar <ProductItemGroup /> correctamente 1`] = `
<Fragment>
  <h2>
    Juegos
  </h2>
  <hr />
  <div
    className="row row-cols-1 row-cols-md-4 g-4"
  >
    <div
      className="col"
      key="1222"
    >
      <ProductItem
        addCart={[MockFunction]}
        item={
          Object {
            "id": "1222",
            "platform": Object {
              "name": "Nintendo 64",
            },
            "price": 23,
            "product": Object {
              "img": "",
              "name": "Super Mario",
              "rating": Object {
                "count": 340,
                "value": 1,
              },
            },
          }
        }
        selectMoney="EUR"
        showDesc={true}
        showDetails={[MockFunction]}
      />
    </div>
    <div
      className="col"
      key="123"
    >
      <ProductItem
        addCart={[MockFunction]}
        item={
          Object {
            "id": "123",
            "platform": Object {
              "name": "Playstation",
            },
            "price": 56,
            "product": Object {
              "img": "",
              "name": "Tomb Raider",
              "rating": Object {
                "count": 340,
                "value": 1,
              },
            },
          }
        }
        selectMoney="EUR"
        showDesc={true}
        showDetails={[MockFunction]}
      />
    </div>
  </div>
  <hr />
</Fragment>
`;
